consensus/polybft/state.go

```diff
@@ -51,123 +46,342 @@ const (
-}
-
-// newStateSyncEvent creates an instance of pending state sync event.
-func newStateSyncEvent(
-	id uint64,
-	sender ethgo.Address,
-	target ethgo.Address,
-	data []byte, log *ethgo.Log,
-) *StateSyncEvent {
-	return &StateSyncEvent{
-		ID:     id,
-		Sender: sender,
-		Target: target,
-		Data:   data,
-		Log:    log,
-	}
-}
-
-func decodeEvent(log *ethgo.Log) (*StateSyncEvent, error) {
-	raw, err := stateTransferEvent.ParseLog(log)
-	if err != nil {
-		return nil, err
+var (
+	// bucket to store rootchain bridge events
+	syncStateEventsBucket = []byte(stateSyncTable)
+	//bucket to store commitments
+	commitmentsBucket = []byte(commitmentTable)
+	// bucket to store message votes (signatures)
+	messageVotesBucket = []byte(messageVoteTable)
+	// bucket to store validator snapshots
+	validatorSnapshotsBucket = []byte(validatorSnapshotTable)
+	// array of all parent buckets
+	parentBuckets = [][]byte{syncStateEventsBucket, commitmentsBucket, messageVotesBucket, validatorSnapshotsBucket}
+	// errNotEnoughStateSyncs error message
+	errNotEnoughStateSyncs = errors.New("there is either a gap or not enough sync events")
+	// errCommitmentNotBuilt error message
+	errCommitmentNotBuilt = errors.New("there is no built commitment to register")
+)
+
+// MemDBIterator is an interface implemented by every custom memdb iterator
+type MemDBIterator interface {
+	Next() (MemDBRecord, error)
+}
+
+// memStateSchema represents the schema of the in-memory db
+var memStateSchema = &memdb.DBSchema{
+	Tables: map[string]*memdb.TableSchema{
+		// list of state syncs
+		stateSyncTable: {
+			Name: stateSyncTable,
+			Indexes: map[string]*memdb.IndexSchema{
+				"id": {
+					Name:    "id",
+					Unique:  true,
+					Indexer: &memdb.UintFieldIndex{Field: "ID"},
+				},
+			},
+		},
+		commitmentTable: {
+			Name: commitmentTable,
+			Indexes: map[string]*memdb.IndexSchema{
+				"id": {
+					Name:    "id",
+					Unique:  true,
+					Indexer: &memdb.UintFieldIndex{Field: "ToIndex"},
+				},
+			},
+		},
+		messageVoteTable: {
+			Name: messageVoteTable,
+			Indexes: map[string]*memdb.IndexSchema{
+				"id": {
+					Name:    "id",
+					Unique:  true,
+					Indexer: &memdb.UintFieldIndex{Field: "Epoch"},
+				},
+				"hash": {
+					Name:    "hash",
+					Unique:  true,
+					Indexer: &memdb.StringFieldIndex{Field: "Hash"},
+				},
+			},
+		},
+		validatorSnapshotTable: {
+			Name: validatorSnapshotTable,
+			Indexes: map[string]*memdb.IndexSchema{
+				"id": {
+					Name:    "id",
+					Unique:  true,
+					Indexer: &memdb.UintFieldIndex{Field: "Epoch"},
+				},
+			},
+		},
+	},
+}
+
+func getUpperBoundUint64(upperBound interface{}) uint64 {
+	if upperBound == nil {
+		// if there is no upper bound provided, than we return everything
+		return math.MaxUint64
 	}
 
-	id, ok := raw["id"].(*big.Int)
+	upper, ok := upperBound.(uint64)
+
 	if !ok {
-		return nil, fmt.Errorf("failed to decode id field of log: %+v", log)
+		panic("state sync table upper bound not a uint64")
```

Comment:
Why do we need panic there?

Reply:
I'll default it to math.MaxUint64, as in case if upper bound was not provided.